File: app/ledger/audit_ledger.py
import uuid
import hashlib
import json
import logging
from typing import Dict, Any, Optional
from datetime import datetime
from sqlalchemy import select, desc
from app.models.database import AsyncSessionLocal
from app.models.ledger_models import FinancialLedger
from app.schemas.ledger import LedgerEntryCreate, LedgerEntryResponse

logger = logging.getLogger(__name__)
GENESIS_HASH = "0000000000000000000000000000000000000000000000000000000000000000"


class AuditLedgerService:
    """
    Append-Only Cryptographic Financial Audit Ledger.
    Ensures absolute data integrity and auditability by linking every financial record
    via SHA-256 hash chaining.
    """

    @classmethod
    def calculate_hash(
        cls,
        previous_hash: str,
        entry_id: str,
        case_id: str,
        entry_type: str,
        amount: float,
        net_recovered: float,
        timestamp_str: str
    ) -> str:
        clean_net = 0.0 if abs(net_recovered) < 1e-9 else float(net_recovered)
        clean_amt = 0.0 if abs(amount) < 1e-9 else float(amount)
        payload = f"{previous_hash}|{entry_id}|{case_id}|{entry_type}|{clean_amt:.4f}|{clean_net:.4f}|{timestamp_str}"
        return hashlib.sha256(payload.encode("utf-8")).hexdigest()

    @classmethod
    async def record_entry(cls, entry: LedgerEntryCreate) -> LedgerEntryResponse:
        entry_id = f"led_{uuid.uuid4().hex[:12]}"
        now = datetime.utcnow()
        timestamp_str = now.strftime("%Y-%m-%d %H:%M:%S")

        # Compute net recovered
        total_costs = entry.gateway_fee + entry.communication_cost + entry.ai_token_cost
        net_recovered = entry.gross_recovered - total_costs if entry.gross_recovered > 0 else -total_costs
        if abs(net_recovered) < 1e-9:
            net_recovered = 0.0

        async with AsyncSessionLocal() as session:
            # 1. Fetch latest previous ledger entry to get hash
            stmt = select(FinancialLedger).order_by(desc(FinancialLedger.sequence_id)).limit(1)
            last_record = (await session.execute(stmt)).scalar_one_or_none()

            previous_hash = last_record.current_hash if last_record else GENESIS_HASH

            # 2. Compute current record SHA-256 hash
            current_hash = cls.calculate_hash(
                previous_hash=previous_hash,
                entry_id=entry_id,
                case_id=entry.case_id,
                entry_type=entry.entry_type.value if hasattr(entry.entry_type, "value") else str(entry.entry_type),
                amount=entry.amount,
                net_recovered=net_recovered,
                timestamp_str=timestamp_str
            )

            # 3. Create ORM instance
            ledger_record = FinancialLedger(
                entry_id=entry_id,
                case_id=entry.case_id,
                entry_type=entry.entry_type,
                amount=entry.amount,
                currency=entry.currency,
                gross_recovered=entry.gross_recovered,
                gateway_fee=entry.gateway_fee,
                communication_cost=entry.communication_cost,
                ai_token_cost=entry.ai_token_cost,
                net_recovered=net_recovered,
                previous_hash=previous_hash,
                current_hash=current_hash,
                metadata_json=entry.metadata,
                timestamp=now
            )

            session.add(ledger_record)
            await session.commit()
            await session.refresh(ledger_record)

            logger.info(f"Recorded Audit Ledger Entry Seq #{ledger_record.sequence_id} (Hash: {current_hash[:12]}...)")

            return LedgerEntryResponse.model_validate(ledger_record)

    @classmethod
    async def verify_chain_integrity(cls) -> Dict[str, Any]:
        """Validates the complete SHA-256 cryptographic chain across all historical ledger records."""
        async with AsyncSessionLocal() as session:
            stmt = select(FinancialLedger).order_by(FinancialLedger.sequence_id.asc())
            records = (await session.execute(stmt)).scalars().all()

            expected_prev_hash = GENESIS_HASH
            for idx, r in enumerate(records):
                if r.previous_hash != expected_prev_hash:
                    return {
                        "is_valid": False,
                        "broken_sequence_id": r.sequence_id,
                        "error": f"Previous hash mismatch at sequence {r.sequence_id}"
                    }

                # Verify current hash recomputation
                timestamp_str = r.timestamp.strftime("%Y-%m-%d %H:%M:%S") if r.timestamp else ""
                recalculated = cls.calculate_hash(
                    previous_hash=r.previous_hash,
                    entry_id=r.entry_id,
                    case_id=r.case_id,
                    entry_type=r.entry_type.value if hasattr(r.entry_type, "value") else str(r.entry_type),
                    amount=r.amount,
                    net_recovered=r.net_recovered,
                    timestamp_str=timestamp_str
                )

                if recalculated != r.current_hash:
                    return {
                        "is_valid": False,
                        "tampered_sequence_id": r.sequence_id,
                        "error": f"Data tampering detected at sequence {r.sequence_id}"
                    }

                expected_prev_hash = r.current_hash

            return {
                "is_valid": True,
                "total_records_verified": len(records),
                "last_verified_hash": expected_prev_hash
            }

File: src/engine/auditLedger.ts
import crypto from "crypto";
import {
  CohortRecoveryStats,
  FinancialLedgerEntry,
  FinancialMetricsResponse,
  LedgerEntryType,
} from "../types.js";

const GENESIS_HASH = "0000000000000000000000000000000000000000000000000000000000000000";

const ledgerRecords: FinancialLedgerEntry[] = [];
let nextSequenceId = 1;

export class AuditLedgerService {
  static calculateHash(params: {
    previous_hash: string;
    entry_id: string;
    case_id: string;
    entry_type: string;
    amount: number;
    net_recovered: number;
    timestamp_str: string;
  }): string {
    const {
      previous_hash,
      entry_id,
      case_id,
      entry_type,
      amount,
      net_recovered,
      timestamp_str,
    } = params;
    const cleanNet = Math.abs(net_recovered) < 1e-9 ? 0.0 : net_recovered;
    const cleanAmt = Math.abs(amount) < 1e-9 ? 0.0 : amount;
    const payload = `${previous_hash}|${entry_id}|${case_id}|${entry_type}|${cleanAmt.toFixed(4)}|${cleanNet.toFixed(4)}|${timestamp_str}`;
    return crypto.createHash("sha256").update(payload).digest("hex");
  }

  static async recordEntry(entry: {
    case_id: string;
    entry_type: LedgerEntryType;
    amount: number;
    currency: string;
    gross_recovered?: number;
    gateway_fee?: number;
    communication_cost?: number;
    ai_token_cost?: number;
    metadata?: Record<string, any>;
  }): Promise<FinancialLedgerEntry> {
    const entryId = `led_${Math.random().toString(36).substring(2, 14)}`;
    const now = new Date();
    // Format YYYY-MM-DD HH:MM:SS
    const timestampStr = now.toISOString().replace("T", " ").substring(0, 19);

    const grossRecovered = entry.gross_recovered || 0.0;
    const gatewayFee = entry.gateway_fee || 0.0;
    const commCost = entry.communication_cost || 0.0;
    const aiCost = entry.ai_token_cost || 0.0;
    const totalCosts = gatewayFee + commCost + aiCost;
    let netRecovered = grossRecovered > 0 ? grossRecovered - totalCosts : -totalCosts;
    if (Math.abs(netRecovered) < 1e-9) {
      netRecovered = 0.0;
    }

    const previousRecord = ledgerRecords.length > 0 ? ledgerRecords[ledgerRecords.length - 1] : null;
    const previousHash = previousRecord ? previousRecord.current_hash : GENESIS_HASH;

    const currentHash = this.calculateHash({
      previous_hash: previousHash,
      entry_id: entryId,
      case_id: entry.case_id,
      entry_type: String(entry.entry_type),
      amount: entry.amount,
      net_recovered: netRecovered,
      timestamp_str: timestampStr,
    });

    const record: FinancialLedgerEntry = {
      sequence_id: nextSequenceId++,
      entry_id: entryId,
      case_id: entry.case_id,
      entry_type: entry.entry_type,
      amount: entry.amount,
      currency: entry.currency,
      gross_recovered: grossRecovered,
      gateway_fee: gatewayFee,
      communication_cost: commCost,
      ai_token_cost: aiCost,
      net_recovered: netRecovered,
      previous_hash: previousHash,
      current_hash: currentHash,
      metadata: entry.metadata || {},
      timestamp: now.toISOString(),
    };

    ledgerRecords.push(record);
    return record;
  }

  static getEntries(limit: number = 50): FinancialLedgerEntry[] {
    return [...ledgerRecords].reverse().slice(0, limit);
  }

  static async verifyChainIntegrity(): Promise<{
    is_valid: boolean;
    total_records_verified: number;
    last_verified_hash?: string;
    broken_sequence_id?: number;
    tampered_sequence_id?: number;
    error?: string;
  }> {
    let expectedPrevHash = GENESIS_HASH;

    for (const r of ledgerRecords) {
      if (r.previous_hash !== expectedPrevHash) {
        return {
          is_valid: false,
          total_records_verified: ledgerRecords.length,
          broken_sequence_id: r.sequence_id,
          error: `Previous hash mismatch at sequence ${r.sequence_id}`,
        };
      }

      const timestampStr = r.timestamp.replace("T", " ").substring(0, 19);
      const recalculated = this.calculateHash({
        previous_hash: r.previous_hash,
        entry_id: r.entry_id,
        case_id: r.case_id,
        entry_type: String(r.entry_type),
        amount: r.amount,
        net_recovered: r.net_recovered,
        timestamp_str: timestampStr,
      });

      if (recalculated !== r.current_hash) {
        return {
          is_valid: false,
          total_records_verified: ledgerRecords.length,
          tampered_sequence_id: r.sequence_id,
          error: `Data tampering detected at sequence ${r.sequence_id}`,
        };
      }

      expectedPrevHash = r.current_hash;
    }

    return {
      is_valid: true,
      total_records_verified: ledgerRecords.length,
      last_verified_hash: expectedPrevHash,
    };
  }

  static getSummaryMetrics(): FinancialMetricsResponse {
    let totalAtRisk = 0.0;
    let totalGrossRecovered = 0.0;
    let totalGatewayFee = 0.0;
    let totalCommCost = 0.0;
    let totalAiCost = 0.0;
    const caseIds = new Set<string>();
    let guardrailBlocked = 0;

    for (const r of ledgerRecords) {
      totalAtRisk += r.amount;
      totalGrossRecovered += r.gross_recovered;
      totalGatewayFee += r.gateway_fee;
      totalCommCost += r.communication_cost;
      totalAiCost += r.ai_token_cost;
      caseIds.add(r.case_id);
      if (r.metadata?.status === "guardrail_blocked" || r.metadata?.status === "paused_dispute") {
        guardrailBlocked++;
      }
    }

    const totalCosts = totalGatewayFee + totalCommCost + totalAiCost;
    const netRecovered = totalGrossRecovered - totalCosts;
    const recoveryRate = totalAtRisk > 0 ? (totalGrossRecovered / totalAtRisk) * 100.0 : 0.0;
    const roiMultiple = totalCosts > 0 ? netRecovered / totalCosts : 0.0;

    return {
      total_revenue_at_risk: Math.round(totalAtRisk * 100) / 100,
      total_recovered_revenue: Math.round(totalGrossRecovered * 100) / 100,
      total_recovery_costs: Math.round(totalCosts * 100) / 100,
      net_recovered_revenue: Math.round(netRecovered * 100) / 100,
      overall_recovery_rate_pct: Math.round(recoveryRate * 10) / 10,
      net_roi_multiple: Math.round(roiMultiple * 100) / 100,
      total_cases_processed: caseIds.size,
      resolved_cases_count: caseIds.size - guardrailBlocked,
      active_cases_count: 0,
      guardrail_blocked_count: guardrailBlocked,
    };
  }

  static getCohortBreakdown(): CohortRecoveryStats[] {
    const cohortMap: Record<string, { cases: number; at_risk: number; recovered: number }> = {
      soft_decline: { cases: 0, at_risk: 0.0, recovered: 0.0 },
      abandoned_checkout: { cases: 0, at_risk: 0.0, recovered: 0.0 },
      overdue_invoice: { cases: 0, at_risk: 0.0, recovered: 0.0 },
      expired_mandate: { cases: 0, at_risk: 0.0, recovered: 0.0 },
    };

    for (const r of ledgerRecords) {
      const rtype = r.metadata?.risk_type || "soft_decline";
      if (cohortMap[rtype]) {
        cohortMap[rtype].cases++;
        cohortMap[rtype].at_risk += r.amount;
        cohortMap[rtype].recovered += r.gross_recovered;
      }
    }

    return Object.entries(cohortMap).map(([rtype, data]) => {
      const atRisk = data.at_risk;
      const rec = data.recovered;
      const rate = atRisk > 0 ? (rec / atRisk) * 100.0 : 0.0;
      return {
        risk_type: rtype,
        total_cases: data.cases,
        total_at_risk_usd: Math.round(atRisk * 100) / 100,
        recovered_usd: Math.round(rec * 100) / 100,
        recovery_rate_pct: Math.round(rate * 10) / 10,
        avg_hours_to_recover: 18.5,
      };
    });
  }
}
